Allow to pass array values to `.in_order_of`

doits opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), the motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

This Pull Request has been created because I wanted to group records with `.in_order_of`. I makes it possible to execute a query like this:

```rb
Posts.in_order_of(:state, [[:published, :canceled], :archived]).order(created_at: :desc)
# =>
# SELECT "posts".* FROM "posts" WHERE "posts"."state" IN (1, 2, 3)
# ORDER BY CASE
#   WHEN "posts"."state" IN (1, 2) THEN 1
#   WHEN "posts"."state" = 3 THEN 2
#  END ASC, "posts"."created_at" DESC
```

This makes it possible to group records together and then order those groups again with another order clause.

### Additional information

I've touched this code for the first time, so I'm not sure if this is the best way to implement it. For example one could use `Array.wrap` instead of `values.is_a?(Array)` checks (I feel this would reduce the performance, no benchmarks yet though).

Another thing this changes is that array values or now not typecasted as a whole but each item. I'm not sure if this is OK or if there might be a case where an array value as a whole should be used for the type cast (e.g. for array typed columns).

Maybe there's even a way I've not come up yet to archive the same result with existing methods.

Feel free to request changes or hint how this can be implemented more efficiently – if you thing this is something to implement at all.

Thanks for reviewing it!

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

MatheusRich commented:
@doits looks like tests are still failing

doits commented:
Thanks for the ping! I was caught off guard by different handling of `NULL` values by different databases – fixed it by explicitly specifying how to treat them (`NULL LAST`).

Rebased on `main`, too.

Now only [one unrelated/flaky test](https://buildkite.com/rails/rails/builds/111297) failed. Cannot retry it in the webinterface unfortunately.

rafaelfranca reviewed (changes requested):
Can you update the documentation to show this usage?

We should also support on the `Enumerable` version.

doits committed:
Allow to pass array values to `.in_order_of`

Passing arrays allows to group records and order those groups with another query:

```rb
Posts
  .in_order_of(:state, [[:published, :canceled], :archived])
  .order(created_at: :desc)
```

doits commented:
Thanks for coming back to this PR!

> Can you update the documentation to show this usage?

Done, rebased and force pushed all in one commit.

> We should also support on the Enumerable version.

I think this can be archived by a simple `flatten(1)`, which would mean `.in_order_of(:id, [[1, 5], 3])` would be the same as `.in_order_of(:id, [1, 5, 3])`. I've added it.

Three remarks though (one is actually a failing test):

1. Now it reorders individual entries even if the are in the same group though:

   ```rb
   [1, 2, 3, 4, 5].in_order_of(:itself, [1, 2, [5, 4, 3]])
   # => [ 1, 2, 5, 4, 3 ]
   ```

   One *could* expect that it does no change the order of the original entries if they are in the same group:

   ```rb
   [1, 2, 3, 4, 5].in_order_of(:itself, [1, 2, [5, 4, 3]])
   # => [ 1, 2, 3, 4, 5 ]
   ```

   Do you think it should be like the latter case? I can change it, though it might be a little bit more complex.

2. `Enumarable#in_order_of` is used with a different expectation in Rails itself here: https://github.com/rails/rails/blob/2ca26346a563a375277e97a09d879df33682df55/activerecord/lib/active_record/relation/finder_methods.rb#L577

   There it actually wants to compare against an array value, not each individual entry of the array, which makes the tests fail.

3. Another thing is the chaining, which does not apply for the Enumerable

    For the Enumerable version, chaining `in_order_of` calls does not make much sense. E.g. for `[1, 2, 3].in_order_of(...).in_order_of(...)`, the keys of the first `in_order_of` do not matter for the final order at all (besides filtering). So the reason why I added it to the active record case (the chaining of `.order` calls) does not apply for the Enumerable version.

So I'm not sure if the change to the `Enumarable#in_order_of` should be made at all or if those methods are different things, even if they have the same name. If the change should stay, https://github.com/rails/rails/blob/2ca26346a563a375277e97a09d879df33682df55/activerecord/lib/active_record/relation/finder_methods.rb#L577 has to be changed. There could be another parameter to `Enumerable.in_order_of`, maybe `flatten_arrays: false`, which could be passed and would *not* flatten the values?

What do you think?

doits commented:
Thanks for your response. I remove the Enumerable changes. One test is failing, though, but it looks like it has nothing to do with the changes?

doits commented:
@rafaelfranca rebased again and still all green, any chance this could be merged? thank you!

fatkodima commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -694,6 +704,16 @@ def order!(*args) # :nodoc:
     #   #     WHEN "conversations"."status" = 0 THEN 2
     #   #   END ASC
     #
+    # For an array value each individual value of the array is typecasted.
```
Can you remove this change (and the changed line above) and rebase the PR? Or edit the PR's checkmark so maintainers can push into it?

doits replied:
Thanks for the feedback!

Rebased on `main` (there were some updates needed) and removed the comment. Hope all test go through 🤞 

> Or edit the PR's checkmark so maintainers can push into it?

Sorry, can't find the setting in the github user interface here how to enable it, otherwise I would 🤔
